### PR TITLE
 fix bugs, wrap up optimizers, fix moe stuff, fix moe parallel

### PR DESCRIPTION
**Pull request by rakkit**

### CODE REVIEW

File: torchtitan/components/optimizer.py [hunk 1]
```diff
@@ -84,12 +88,12 @@ def _extract_param_groups(
         else None
     )
 
-    if optimizer_config is None or not param_groups_config:
```
Comment (rakkit) [inline torchtitan/components/optimizer.py:87]: `if optimizer_config is None or not param_groups_config:` 
is same as
`if param_groups_config is None:`

it only triggered that if we are not using muon/scion (i.e. we are using adamw).

& we do need to have `param_names` in optimizer -> log_norm,   so we have to plug in `param_names` for case of adamw

File: torchtitan/components/optimizer.py [hunk 2]
```diff
@@ -164,18 +171,21 @@ def __init__(
...
+                extra_kwargs.setdefault("model", model)
             if is_muon or is_scion:
                 kwargs.update(extra_kwargs)
             self.optimizers.append(optimizer_cls(params, **kwargs))
             all_params.extend(params)
         self._validate_length(len(self.model_parts))
         # Do not separately save the external settings in
         # optimizer defaults.
-        del optimizer_kwargs["param_groups"]
+        if "param_groups" in optimizer_kwargs:
```
Comment (rakkit) [inline torchtitan/components/optimizer.py:187]: seems for adamw, for somereason param_groups may not exists in optimizer_kwargs.

anyhow its a sate del now

File: torchtitan/components/optimizer.py [hunk 3]
```diff
@@ -240,15 +254,26 @@ def compute_grad(p, optimizer=None, **kwargs):
...
+            ), "Weight decay not supported for grad computation."
 
             param_optim_state = optimizer.state[p]
-            step = param_optim_state["step"].item()
-            bias_correction1 = 1 - beta1**step
-            bias_correction2 = 1 - beta2**step
-            denom = (param_optim_state["exp_avg_sq"].sqrt() / math.sqrt(bias_correction2)) + eps
-            step_size = 1 / bias_correction1
-            g = step_size * param_optim_state["exp_avg"].div(denom)
+            if "step" not in param_optim_state:
```
Comment (rakkit) [inline torchtitan/components/optimizer.py:262]: In some cases, `optim.State` may not be created.  (for example, EP+Adamw).

File: torchtitan/components/optimizer.py [hunk 4]
```diff
@@ -301,8 +333,28 @@ def get_parameter_norms(self):
                         if "tok_embeddings" in n:
                             p, update = p.T, update.T
                         for norm_name, norm_func in NORM_FUNCTIONS.items():
-                            norms[f"model_part_{i}/{n}/param/{norm_name}"] = norm_func(p)
-                            norms[f"model_part_{i}/{n}/update/{norm_name}"] = norm_func(update)
+                            if norm_name == "rms_to_l1" and p.ndim < 2:
```
Comment (rakkit) [inline torchtitan/components/optimizer.py:336]: for `adam,` `rms_to_l1` (and `l2_to_l1` in principle) will fail in the case that we have a learnable layer/rms norm.    Or any 1D weights.

File: torchtitan/components/optimizer.py [hunk 5]
```diff
@@ -301,8 +333,28 @@ def get_parameter_norms(self):
...
+                                # In case we have lernable Layer/RMSNorm
+                                continue
+                            if norm_name == "spectral" and (
```
Comment (rakkit) [inline torchtitan/components/optimizer.py:339]: This is a wrap for MoE (group expert implementation).
Because the parameters now is [experts, Dim-in, Dim-out]. We shall compute and log the `spectral` for each expert.

File: torchtitan/models/MoEllama/moellama.py [hunk 6]
```diff
@@ -114,18 +134,27 @@ def __init__(
...
         # Cache for accumulated adjustments
         self._accumulated_adjustment = None
         self._num_accumulated = 0
+        self.needs_reduce_bias_update = False
 
     def __repr__(self):
         return f"Gate(experts={self.experts}, topk={self.topk}, bias={self.bias is not None})"
 
+    def init_weights(self):
+        # nn.init.xavier_uniform_(self.expert_embeddings)
```
Comment (rakkit) [inline torchtitan/models/MoEllama/moellama.py:154]: @ofivite

File: torchtitan/models/MoEllama/moellama.py [hunk 7]
```diff
@@ -922,8 +682,47 @@ def _precompute_freqs_cis(self) -> torch.Tensor:
         )
 
     def update_gate_bias(self):
+        gates_to_update = []
```
Comment (rakkit) [inline torchtitan/models/MoEllama/moellama.py:685]: So, we need to gather/reduce to sync the usage of experts.  
rather than do it for `layers` times, we can do it `once` aggregate them here

File: torchtitan/models/MoEllama/moellama.py [hunk 8]
```diff
@@ -922,8 +682,47 @@ def _precompute_freqs_cis(self) -> torch.Tensor:
...
+                ].clone()
+                gate.needs_reduction = False  # Mark as reduced
+                start_idx += experts
+
         for layer in self.layers.values():
-            layer.update_gate_bias()
+            if hasattr(layer, "update_gate_bias"):
+                layer.update_gate_bias()
+
+    def get_sparsity_ratio(self):
```
Comment (rakkit) [inline torchtitan/models/MoEllama/moellama.py:719]: just make parameters count easier

File: torchtitan/models/llama/__init__.py [hunk 9]
```diff
@@ -35,56 +35,127 @@
...
+        norm_eps=1e-12,
+        first_in_init_fn_type="orthogonal",
+        intermediate_init_fn_type="orthogonal",
         final_out_init_fn_type="zeros",
     ),
     # 55M parameters
     "debugmodel-multiplier-2": TransformerModelArgs(
-        dim=512, n_layers=16, n_heads=8, n_kv_heads=None, ffn_dim_multiplier=1, multiple_of=256,
-        rope_theta=500000, depth_init=False, qk_norm=True, norm_eps=1e-12, init_std=0
+        dim=512,
```
Comment (rakkit) [inline torchtitan/models/llama/__init__.py:54]: remove `init_std`

File: torchtitan/tools/utils.py [hunk 10]
```diff
@@ -48,14 +48,33 @@ def collect(reason: str):
 
 
 def get_num_params(model: nn.Module, exclude_embedding: bool = False) -> int:
-    num_params = sum(p.numel() for p in model.parameters())
+    not_ffn_params = 0
```
Comment (rakkit) [inline torchtitan/tools/utils.py:51]: now we can compute the activate/total parameters.

and activate parameters will be used to compute flops per token

**Comment by janEbert** [comment]
Thanks for all the explanations, very helpful!

**Comment by janEbert** [comment]
Merged on branch, will follow up with commit hashes after future clean-up.
